File: tests/mozmill/tests/testLinks/testHTMLAnchorElement/testOpenInNewWindow.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

var rpRootDir = "../../../";
var rpConst = require(rpRootDir + "lib/constants");
var rootDir = rpRootDir + rpConst.mozmillTestsRootDir;

var {assert, expect} = require(rootDir + "lib/assertions");
var prefs = require(rootDir + "lib/prefs");
var tabs = require(rootDir + "firefox/lib/tabs");
var utils = require(rootDir + "lib/utils");

var rpUtils = require(rpRootDir + "lib/rp-utils");

const TEST_URL = "http://www.maindomain.test/link_1.html";


var setupModule = function(aModule) {
  aModule.controller = mozmill.getBrowserController();
  aModule.otherController = null;

  aModule.tabBrowser = new tabs.tabBrowser(aModule.controller);
  aModule.tabBrowser.closeAllTabs();
  aModule.otherTabBrowser = null;

  prefs.setPref(rpConst.PREF_DEFAULT_ALLOW, false);
};

var teardownModule = function(aModule) {
  prefs.clearUserPref(rpConst.PREF_DEFAULT_ALLOW);
  utils.closeContentAreaContextMenu(aModule.controller);
  aModule.tabBrowser.closeAllTabs();

  closeWindow(aModule.otherController);
};

var testOpenInNewWindow = function () {
  controller.open(TEST_URL);
  controller.waitForPageLoad();

  let link = rpUtils.getLink(controller);
  let linkURL = link.getNode().href;

  let i = 1;
  while (true === openNextWindow(i, link)) {
    assert.waitFor(function () {
      // Make sure that we work on the correct window
      var windows = mozmill.utils.getWindows("navigator:browser");
      for (var j = 0; j < windows.length; ++j) {
        if (windows[j] !== controller.window) {
          otherController = new mozmill.controller.MozMillController(windows[j]);
          otherTabBrowser = new tabs.tabBrowser(otherController);
          break;
        }
      }

      return !!otherController;
    }, "Newly opened browser window has been found.");

    rpUtils.waitForTabLoad(controller, otherTabBrowser.getTab(0));

    assert.equal(otherController.tabs.activeTab.location.href, linkURL,
        "The location in the new window is correct.");

    assertNoRedirects();
    closeWindow(otherController);
    otherController = null;
    otherTabBrowser = null;
    ++i;
  }
};

var closeWindow = function(_controller) {
  if (_controller && _controller.window) {
    _controller.window.close();
  }
};



/**
 * Opens the next tab.
 * @return {boolean}
 *         true if a new tab has been opened.
 *         false if no tab needs to open anymore.
 */
var openNextWindow = function(i, link) {
  var contextMenuItemID;

  switch (i) {
    case 1:
      // Context Menu: Open Link in New Window
      contextMenuItemID = "#context-openlink";
      break;

    case 2:
      // Context Menu: Open Link in New Private Window
      contextMenuItemID = "#context-openlinkprivate";
      break;

    default:
      return false;
      break;
  }

  // There seems to be a bug in Mozmill: the right-click is not performed the
  // second time. The workaround seems to be to wait for some miliseconds before
  // continuing. I set the timeout to 100 ms. If this time is still not enough
  // for anybody, simply increase it.
  // It seemed that this bug has to do with the focus. However, the
  // HTMLElement.blur() function doesn't work for <a> Elements.
  controller.sleep(100);

  var contextMenu = controller.getMenu("#contentAreaContextMenu");
  contextMenu.select(contextMenuItemID, link);
  return true;
};

/**
 * Assert that the link clicks have not been detected as redirects.
 */
var assertNoRedirects = function() {
  var tabBrowsers = [tabBrowser, otherTabBrowser];

  for (let i = 0; i < tabBrowsers.length; ++i) {
    var panel = tabBrowsers[i].getTabPanelElement(0,
        '/{"value":"' + rpConst.REDIRECT_NOTIFICATION_VALUE + '"}');
    assert.ok(false === panel.exists(),
        "Following the link didn't cause a redirect");
  }
};

File: tests/marionette/tests/links/text_selection/test_open_in_new_window.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from rp_ui_harness import RequestPolicyTestCase
from rp_puppeteer.ui.redirect_notification import RedirectNotification
from rp_puppeteer.ui.context_menu import ContextMenu
from rp_puppeteer.ui.tabs import Tabs
from rp_puppeteer.ui.web_utils import WebUtils


TEST_URL = "http://www.maindomain.test/link_1.html";
PREF_DEFAULT_ALLOW = "extensions.requestpolicy.defaultPolicy.allow"


class TestOpenInNewWindow(RequestPolicyTestCase):

    def setUp(self):
        RequestPolicyTestCase.setUp(self)
        self.prefs.set_pref(PREF_DEFAULT_ALLOW, False);

        self.redir = RedirectNotification(lambda: self.marionette)
        self.web_utils = WebUtils(lambda: self.marionette)
        self.tabs = Tabs(lambda: self.marionette)
        self.ctx_menu = ContextMenu(lambda: self.marionette)

        self.origin_window = self.windows.current

    def tearDown(self):
        self.windows.close_all(exceptions=[self.origin_window])
        RequestPolicyTestCase.tearDown(self)

    def test_open_in_new_window(self):
        with self.marionette.using_context("content"):
            # load the test url
            self.marionette.navigate(TEST_URL)

            # find the URL string and its wrapping element
            url_wrapper = self.marionette.find_element("id", "text_url_1")
            url = url_wrapper.get_attribute("textContent")

            # select the URL
            self.web_utils.select_element_text(url_wrapper)

        with self.marionette.using_context("chrome"):
            for _ in self.open_sel_url_multi(url_wrapper):
                self.assertEqual(len(self.windows.all), 2,
                                 "A new window has been opened.")

                dest_window_handle = self.windows.focused_chrome_window_handle
                self.assertNotEqual(dest_window_handle, self.origin_window.handle,
                                    "The new window has been focused.")
                dest_window = self.windows.create_window_instance(
                    dest_window_handle
                )

                # Await page load.
                dest_tab = dest_window.tabbar.selected_tab
                self.tabs.wait_until_loaded(dest_tab)

                # Do checks in the destination's window.
                dest_window.switch_to()
                self.assertEqual(dest_tab.location, url,
                                 "The location in the new window is correct.")
                self.assertFalse(self.redir.panel_exists(),
                                 ("Following the link didn't cause a "
                                  "redirect in the destination window."))

                # Close the new window.
                self.windows.close_all(exceptions=[self.origin_window])

                # Do checks in the origin's window.
                self.origin_window.switch_to()
                self.assertFalse(self.redir.panel_exists(),
                                 ("Following the link didn't cause a "
                                  "redirect in the origin window."))

    def open_sel_url_multi(self, url_wrapper):
        """Open a selected URL in new window using different methods."""

        context_menu_ids = [
            "context-openlink",       # Open Link in New Window
            "context-openlinkprivate" # Open Link in New Private Window
        ]

        for id in context_menu_ids:
            self.ctx_menu.select_entry(id, url_wrapper)
            yield
